File: CloudWalk linguagens/java/cloudwalk - java/src/main/java/com/game/Methods.java
package com.game;

import java.io.FileWriter;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import com.github.cliftonlabs.json_simple.JsonObject;

public class Methods{
    static String[] vector;
    static String[] userLine;
    static ArrayList<String> log;

    private static String splitClient(String line){
        userLine = line.split("ClientUserinfoChanged: ");
        String[] player =  userLine[userLine.length -1].split("");
        return player[0];
    }

    private static void includePlayerNumberArray(Integer playerNumber){
        if (!App.playersNumber.contains(playerNumber)){
            App.playersNumber.add(playerNumber);
        }   
    }

    private static void includePlayerNameArray(String name){
        if (!name.isEmpty()){
            if(App.playersName.isEmpty()){
                App.playersName.add(name);
                App.playerKills.put(name, 0);
            }else if (!App.playersName.contains(name)){
                App.playersName.add(name);
                App.playerKills.put(name, 0);
            }
        }
    }

    private static String getName(){
        String name = userLine[userLine.length -1];
        String[] userName = name.split("n\\\\|\\\\t");
        if(!(userName.length <= 1)){
            return userName[1];
        }
        return null;
    }

    private static String splitKillLine(String line){
        String[] status = line.split(":");
        if(!(status.length <= 2)){
            status = status[2].split(" ");
            return status[2];
        }
        return null;
    }

    private static void scoreKill(Integer playerStatus, boolean status){
        int index = App.playersNumber.indexOf(playerStatus);
        String name = App.playersName.get(index);
        int value = (status == true) ? 1 : -1;
        int previousScore = (int)  App.playerKills.get(name); 
        int less_value = previousScore + value;
        App.playerKills.put(name, less_value);
    }

    private static void weaponCountKill(String weapon){
        if (!App.weapons.contains(weapon)){
            App.weapons.add(weapon);
            App.killsByMeans.put(weapon, 1);
        }else{
            int indexWeapon = App.weapons.indexOf(weapon);
            String nameWeapons = App.weapons.get(indexWeapon);
            int previousValue = (int) App.killsByMeans.get(nameWeapons);
            App.killsByMeans.put(nameWeapons, previousValue + 1);
        }
    }

   public static ArrayList<String> splitLogGame(){
        String path = "./log/quake.log";
        try {
            String[] logGames = Files.readString(Paths.get(path)).split("InitGame:");
            log = new ArrayList<>(Arrays.asList(logGames));
            return log;
        } catch (IOException e) {
            e.printStackTrace();
        }
        return null;
    }

    public static String[] splitLine(int i){
        return log.get(i).split("\n");
    }

    public static void gameName(int i){
        String gameName = "game_" + Integer.toString(i);
        App.newJson = new JsonObject();
        App.json.put(gameName, App.newJson);
    }

    public static void gameLogicFilters(String[] gameBlock){
        for(String line : gameBlock){
            if(line.contains("ClientUserinfoChanged:")){
                playerName(line);               
            }else if(line.contains("Kill:") && line.contains("<world>")){
                killPlayer(false, line); 
            }else if(line.contains("Kill:") && !line.contains("<world>")){
                killPlayer(true, line);  
            }
        }
    }

    public static void playerName(String line){
        String playerNumber = splitClient(line);
        if(!playerNumber.isEmpty()){
            includePlayerNumberArray(Integer.parseInt(playerNumber));
            String name = getName();
            includePlayerNameArray(name);
        }
    }

    public static void killPlayer(boolean status, String line){
        String playerStatus = splitKillLine(line);
        String[] weaponLine = line.split(" ");
        String weapon = weaponLine[weaponLine.length -1];
        if(!playerStatus.isEmpty()){
            scoreKill(Integer.parseInt(playerStatus), status);
            App.totalKills += 1;
            weaponCountKill(weapon);
        }
    }

    public static void jsonStructure(JsonObject json){
        json.put("total_kills", App.totalKills);
        json.put("players", App.playersName);
        json.put("kills", App.playerKills);
        json.put("kills_by_means", App.killsByMeans);
    }

    public static void createJsonFile() throws IOException{
        Gson gson = new GsonBuilder().setPrettyPrinting().create();
        // organazingJsonOerde();
        String result = gson.toJson(App.json);
        FileWriter fileWriter = new FileWriter("./results/result.json");
        fileWriter.write(result);
        fileWriter.close();
    }
}

File: CloudWalk linguagens/CloudWalk-challenge/methods.rb
def split_log_game()
    @log = File.read('./log/quake.log').split('InitGame:')
    @log.size
end

def split_line(i)
    @log[i].split("\n")
end

def game_name(i)
    @json["game_#{i.to_s}"] = {}
end

def game_logic_filters(game_block)
    game_block.each do |line|
        player_name(line)        if line.include? 'ClientUserinfoChanged:'
        kill_player(false, line) if line.include? 'Kill:'   && '<world>'
        kill_player(true, line)  if (line.include? 'Kill:') && (!line.include? '<world>')
    end
end

def player_name(line)
    player_number = split_client(line)
    include_player_number_array(player_number)
    name = get_name()
    include_player_name_array(name)
end

def kill_player(status, line)
    player_status = split_kill_line(line)
    weapon = line.split.last 
    score_kill(player_status, status)
    @total_kills += 1
    weapon_count_kill(weapon)
end

def json_structure(json)
    json['total_kills'] = @total_kills
    json['players'] = @players_name
    json['kills'] = @player_kills
    json['kills_by_means'] = @kills_by_means
    json
end

def create_json_file()
    result = @json.to_json
    File.open("./result/result.json", "w") { |f| f.write result }
end

private

def split_client(line)
    @user_line = line.split('ClientUserinfoChanged: ')
    player =  @user_line.last.split
    player[0]
end

def include_player_number_array(player_number)
    if !@players_number.include? player_number
        @players_number << player_number
    end
end

def include_player_name_array(name)
    if !@players_name.include? name
        @players_name << name
        @player_kills[name] = 0
    end
end

def get_name()
    name = @user_line.last
    name = name.split(/n\\|\\t/)
    name[1]
end

def split_kill_line(line)
    status = line.split(':')
    status = status[2].split
    status[1]
end

def score_kill(player_status, status)
    index = @players_number.find_index(player_status)
    name = @players_name[index]
    value = status == true ? 1 : -1
    less_value = @player_kills[name] + value
    @player_kills[name] = less_value
end

def weapon_count_kill(weapon)
    if !@weapons.include? weapon
        @weapons << weapon
        @kills_by_means[weapon] = 1
    else
        index_weapon = @weapons.find_index(weapon)
        index_weapons = @weapons[index_weapon]
        @kills_by_means[index_weapons] += + 1
    end
end